File: DAA/Assign 3 Fractional knapsack.py
def fractionalKnapsack(items, capacity):
    for item in items:
        item['value_per_weight'] = item['value'] / item['weight']

    items.sort(key=lambda x: x['value_per_weight'], reverse=True)

    total_value = 0
    knapsack = []

    for item in items:
        if item['weight'] <= capacity:
            total_value += item['value']
            knapsack.append(item)
            capacity -= item['weight']
        else:
            fraction = capacity / item['weight']
            total_value += item['value'] * fraction
            knapsack.append({'name': item['name'], 'value': item['value'] * fraction, 'weight': item['weight'] * fraction})
            break

    return total_value, knapsack

# Input from the user
n = int(input("Enter the number of items: "))
items = []
for i in range(n):
    name = input(f"Enter the name of item {i + 1}: ")
    value = float(input(f"Enter the value of item {i + 1}: "))
    weight = float(input(f"Enter the weight of item {i + 1}: "))
    items.append({'name': name, 'value': value, 'weight': weight})

capacity = float(input("Enter the knapsack capacity: "))

max_value, selected_items = fractionalKnapsack(items, capacity)
print("Maximum value that can be obtained:", max_value)
print("Selected items:", selected_items)

File: DAA C++/DAA C++/Assign 3 fractional knapsack.cpp
#include<iostream>
#include<vector>
#include<Algorithm>
using namespace std;

struct Item{
    int weight;
    int value;
    double value_per_weight;

};

bool compare(Item a, Item b){
    return a.value_per_weight >b.value_per_weight;
}

double fractionalknapsack(vector<Item>&items,int capacity){
    sort(items.begin(),items.end(),compare);

    double totalValue=0.0;
    for(const Item &item :items){
        if(capacity==0){
            break;
        }

        double amountTaken=min(item.weight,capacity);
        totalValue+=(amountTaken*item.value_per_weight);
        capacity-=amountTaken;
    }
    return totalValue;
}

int main(){
    int n,capacity;
    cout<<"Enter items"<<endl;
    cin>>n;
    cout<<"Enter bag Capacity"<<endl;
    cin>>capacity;

    vector<Item>items(n);

    cout<<"Enter Weight and Value of each item "<<endl;
    for(int i=0;i<n;i++){
        cin>>items[i].weight>>items[i].value;
        items[i].value_per_weight=static_cast<double>(items[i].value)/items[i].weight;
    }

    double maxValue =fractionalknapsack(items,capacity);
    cout<<"Max value is "<<maxValue;

    return 0;


}

